### ISSUE TITLE
Strive for 100% line coverage for java grammar

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#2973]
Since we added code coverage on the Java grammar in PR #2970 ( [pom.xml#L1800-L1809](https://github.com/checkstyle/checkstyle/blob/6bcc84946bd9a463b56774a2f86f5e45c40c1ee5/pom.xml#L1800-L1809)), we should try to increase the Lexer and the Recognizer's coverage to 100% line coverage so we know we are covering it all.
##

**Comment by romani** [comment]
@rnveach , we need to revert "removed unused FLOAT_SUFFIX from java.g" and do hotfix release
as we damaged compatibility by Tokens .

**Comment by rnveach** [comment]
I am unassigning myself. Line coverage is high 96 for both files.
I'm not totally sure how much more can be finished. Antlr generates alot of dead code, and a few remaining places are that, either dead zones or error protections.
We should concentrate more on valid cases then antlr's own error protection.
@romani If you wish to close this, I am fine.

**Comment by romani** [comment]
I need to review this, as it is completely ok for code generators to generate over cautious code. But we should make notes to all/most of non-covered cases to keep log of them. It might not be a surprise if we in few years change our mind about this dead code.

@rnveach , as you were already in that code for a while, could you please make a notes here about each dead code.
http://checkstyle.sourceforge.net/cobertura/com.puppycrawl.tools.checkstyle.grammars.GeneratedJavaLexer.html
http://checkstyle.sourceforge.net/cobertura/com.puppycrawl.tools.checkstyle.grammars.GeneratedJavaRecognizer.html

**Comment by rnveach** [comment]
GeneratedJavaLexer
````
 1080 else {
 1081    throw new NoViableAltForCharException((char)LA(1), getFilename(), getLine(), getColumn());
 1082 }
````
Dead because top case is checking for `\r` and `\n` and this else is checking if it is none of those.
I believe there are a some other dead spots similar to this like:
````
 3697 default:
 3698 {
 3699     throw new NoViableAltForCharException((char)LA(1), getFilename(), getLine(), getColumn());
 3700 }
````

Nothing else is jumping out at me. I think the lexer has dead code and the recognizer just has alot of error protections.

**Comment by rnveach** [comment]
@nmancus1 Can you confirm that either having the Visitor at 100% line coverage or 100% mutation coverage (or both) should give us enough coverage for what basically amounts to the same as covering the grammar?

In the end, we just want to make sure new grammar is covered for any change we may make to it. We had no way before except to use code coverage before mutation and the visitor. I can understand not covering weird exception handling, unfortunately that is what is included in code coverage. This isn't directly our code, but it is generated from our grammar.

Maybe you could test out a recent grammar change by removing the visitor and test changes and leave in the grammar changes and see how both percentages in the visitor are affected. Current tests would have to remain functional with an incomplete visitor in this case.

If everything in the visitor covers us and the grammar indirectly, then I agree to closing this issue and not needing another one. If we are not covered, I am open to other ideas to show the grammar is covered without diving into exception handling.

**Comment by nmancus1** [comment]
> In the end, we just want to make sure new grammar is automatically covered for any change we may make to it by CI.

If a rule is added to the grammar, but not overridden in the visitor, this means that we are not building an AST in this rule; therefore the rule is not covered. If this happens, then `testAllVisitMethodsAreOverridden` will catch it:

```bash


```

**Comment by rnveach** [comment]
@romani I am fine to close this since it sounds like we will be forced to ensure we continue to have coverage for future changes, which is what really matters.
